linera-core/src/client/mod.rs

```diff
@@ -1203,143 +1159,144 @@ where
+                                block_range = ?block_batch_range,
+                                received_num = hashes.len(),
+                                "Validator sent invalid number of certificate hashes."
+                            );
+                            return Err(NodeError::InvalidChainInfoResponse);
+                        }
+                        Ok(hashes)
+                    })
+                }),
+        )
+        .await?
+        .into_iter()
+        .flatten()
+        .collect();
 
-            for certificate in remote_certificates {
-                match self
-                    .check_certificate(max_epoch, &committees, &certificate)
-                    .await?
-                {
-                    HandleCertificateResult::FutureEpoch => {
-                        warn!(
-                            "Postponing received certificate from {:.8} at height {} \
-                            from future epoch {}",
-                            chain_id,
-                            certificate.value().height(),
-                            certificate.value().epoch()
-                        );
-                        // Stop the synchronization here. Do not increment the tracker further so
-                        // that this certificate can still be downloaded later, once our committee
-                        // is updated.
-                    }
-                    HandleCertificateResult::OldEpoch => {
-                        tracker.push(chain_id, certificate.value().height());
-                        // This epoch is not recognized any more. Let's skip the certificate.
-                        // If a higher block with a recognized epoch comes up later from the
-                        // same chain, the call to `receive_certificate` below will download
-                        // the skipped certificate again.
-                        warn!(
-                            "Skipping received certificate from past epoch {:?}",
-                            certificate.value().epoch()
-                        );
-                    }
-                    HandleCertificateResult::New => {
-                        tracker.push(chain_id, certificate.value().height());
-                        certificates.push(certificate);
-                    }
+        // Download the block certificates.
+        let remote_certificates = remote_node
+            .node
+            .download_certificates(certificate_hashes)
+            .await?;
+
+        // Check the signatures and keep only the ones that are valid.
+        let mut certificates = Vec::new();
+        for certificate in remote_certificates {
+            let sender_chain_id = certificate.value().chain_id();
+            let height = certificate.value().height();
+            let epoch = certificate.value().epoch();
+            match self
+                .check_certificate(max_epoch, &committees, &certificate)
```

Comment:
I didn't notice it before (maybe something have changed in the meantime) but it looks like there's no need for `check_certificate` to be `async`?

Reply:
We could try rewriting that w/o the `let mut certificates` and the `for`-loop (to avoid new allocations) w/ `filter_map`.

Reply:
Good point, I'll make it `sync`! Using iterators is a bit ugly, because we'd return a `Result<Option<Certificate>, _>`, so we'd have to do `collect()?.into_iter().flatten().collect()`.